rts/lra/jaxrs/src/main/java/io/narayana/lra/filter/ServerLRAFilter.java

```diff
@@ -460,46 +575,149 @@ private boolean isJaxRsCancel(ContainerRequestContext requestContext, ContainerR
     }
 
-    private URI startLRA(URI parentLRA, Method method, Long timeout) {
-        // timeout should already have been converted to milliseconds
-        String clientId = method.getDeclaringClass().getName() + "#" + method.getName();
+    // the request filter may perform multiple and in failure scenarios the LRA may be left in an ambiguous state:
+    // the following structure is used to track progress so that such failures can be reported in the response
+    // filter processing
+    private enum ProgressStep {
+        Left ("leave succeeded"),
+        LeaveFailed("leave failed"),
+        Started("start ok"),
+        StartFailed("start failed"),
+        Joined("join succeeded"),
+        JoinFailed("join failed"),
+        Ended("end ok"),
+        CloseFailed("close failed"),
+        CancelFailed("cancel failed");
+
+        final String status;
+
+        ProgressStep(final String status) {
+            this.status = status;
+        }
 
-        return lraClient.startLRA(parentLRA, clientId, timeout, ChronoUnit.MILLIS);
+        @Override
+        public String toString() {
+            return status;
+        }
     }
 
-    private void resumeTransaction(URI lraId) {
-        // nothing to do
+    // list of steps (both successful and unsuccesful) performed so far by the request and response filter
+    // and is used for error reporting
+    private static class Progress {
+        static EnumSet<ProgressStep> failures = EnumSet.of(
+                ProgressStep.LeaveFailed,
+                ProgressStep.StartFailed,
+                ProgressStep.JoinFailed,
+                ProgressStep.CloseFailed,
+                ProgressStep.CancelFailed);
+
+        ProgressStep progress;
+        String reason;
+
+        public Progress(ProgressStep progress, String reason) {
+            this.progress = progress;
+            this.reason = reason;
+        }
+
+        public boolean ok() {
+            return !failures.contains(progress);
+        }
     }
 
-    private String getCompensatorId(URI lraId, UriInfo uriInfo, Long timeout) {
-        Map<String, String> terminateURIs = NarayanaLRAClient.getTerminationUris(resourceInfo.getResourceClass(), uriInfo, timeout);
+    // convert the list of steps carried out by the filters into a warning message
+    private String processLRAOperationFailures(ArrayList<Progress> progress) {
+        StringJoiner badOps = new StringJoiner(", ");
+        StringJoiner goodOps = new StringJoiner(", ");
+        StringBuffer code = new StringBuffer("-");
 
-        if (!terminateURIs.containsKey("Link")) {
-            throwGenericLRAException(lraId, Response.Status.BAD_REQUEST.getStatusCode(),
-                    "Missing complete or compensate annotations");
-        }
+        progress.forEach(p -> {
+            if (p.ok()) {
+                code.insert(0, p.progress.ordinal());
+                goodOps.add(String.format("%s (%s)", p.progress.name(), p.progress.status));
```

Comment:
Why the `ProgressStep#status` is reported for success and not for the failure?

Reply:
Because the stuff in the brackets is only there to give more information about the step.
The failure branch includes the exception message whereas in the successful branch it adds nothing other than a more verbose description of name().